### PR TITLE
[Connector API] Implementation of list, get, delete connector endpoints

### PR DESCRIPTION
**Pull request by jedrazb**
## [Connector API] Implementation of list, get, delete connector endpoints

- Follow up to #101884
- Add new endpoints for Connector API:
  - list
  - get
  - delete
- Add parsers for data structures used to represent `Connector` document
  - Parsers will be used for parsing user-provided data in update endpoints implemented in futures PRs. They also act as verification of the payload shape and data (e.g. if `Instant` or `Cron` expression are misformatted we will throw an exception)
  - Parsers are used for `get` and `list` endpoints to construct connector object, this allows for subsequent manipulation on the instance data (e.g. we want to support obfuscation of client secrets stored in index with `*****`)

### Verification
- BWC tests
- Yaml tests
- Unit tests
- Manual tests

Labelling PR as `>non-issue` since new endpoints are hidden under a feature flag for now.

### CODE REVIEW

**Review by kderusso — COMMENTED** [review]
Completed a first pass review @jedrazb - nice work so far. I've added some questions and some requests for tests, looking forward to iterating on it. 

One nitpick for future reference: Adding API calls one at a time will make for smaller PRs with hopefully faster review turnaround. :)

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/action/GetConnectorActionResponseBWCSerializingTests.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.connector.action;
+
+public class GetConnectorActionResponseBWCSerializingTests {
+
+    // private BytesReference connector;
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/action/GetConnectorActionResponseBWCSerializingTests.java:12]: My bad, I completely missed that I haven't actually implemented it 👍

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/action/ListConnectorActionResponseBWCSerializingTests.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.connector.action;
+
+public class ListConnectorActionResponseBWCSerializingTests {}
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/action/ListConnectorActionResponseBWCSerializingTests.java:10]: Yes, my bad - I intended to implement this before marking this as ready for review. Will fix :)

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorTestUtils.java [hunk 3]
```diff
@@ -36,6 +37,18 @@ public static PutConnectorAction.Request getRandomPutConnectorActionRequest() {
         );
     }
 
+    public static PutConnectorAction.Request getRandomDeleteConnectorActionRequest() {
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorTestUtils.java:40]: Good catch, that was unused and should be removed

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFeatures.java [hunk 4]
```diff
@@ -231,9 +205,156 @@ public ConnectorFeatures build() {
...
                 filteringRules,
                 incrementalSyncEnabled,
-                syncRulesAdvancedEnabled,
-                syncRulesBasicEnabled
+                syncRulesFeatures
             );
         }
     }
+
+    public static class FeatureEnabled implements ToXContentObject, Writeable {
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFeatures.java:213]: This was to deal with some technical debt in `Connector` representation of features in status quo. 

Some features are represented in the `features` object as:
```
"my_feauture": true
```

Whereas some are represented as:
```
"my_feature": {"enabled": true}
```

`FeatureEnabled` is helper class that helps solely with serialising (`toXContent`) and deserialising (`parse`) logic for the nested example. 

I felt that hacking the parser of `ConnnectorFeatures` class would be a bit ugly. I will add a comment explaining why we need the `FeatureEnabled`  class.

I double-checked the logic and I will change `Boolean` to `boolean` because the `enabled` property should not be nullable.

File: rest-api-spec/src/main/resources/rest-api-spec/api/connector.delete.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
+{
+  "connector.delete": {
+    "documentation": {
+      "url": "https://www.elastic.co/guide/en/enterprise-search/current/connectors.html",
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply rest-api-spec/src/main/resources/rest-api-spec/api/connector.delete.json:4]: Yup that URL is a placeholder, I tried excluding the `"url": "...."` from the json spec but it causes tests to fail. The relevant docs will be added and this link will be update before the feature flag is removed. :)

File: rest-api-spec/src/main/resources/rest-api-spec/api/connector.delete.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      "description": "Deletes a connector."
+    },
+    "stability": "experimental",
+    "visibility": "public",
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply rest-api-spec/src/main/resources/rest-api-spec/api/connector.delete.json:8]: Thank you for flagging (pun intended), will fix

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFiltering.java [hunk 7]
```diff
@@ -106,124 +148,207 @@ public ConnectorFiltering build() {
...
+                return this;
+            }
+
+            public FilteringRules build() {
+                return new FilteringRules(advancedSnippet, rules, filteringValidationInfo);
+            }
+        }
+    }
+
+    public static class AdvancedSnippet implements Writeable, ToXContentObject {
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorFiltering.java:262]: Yeah, the ConnectorFiltering class grew too big. I will refactor all filtering related classes into new package to keep this organized, will add javadocs as well

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/300_connector_put.yml
[the remark this answers is not in the document]
Comment (jedrazb) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/300_connector_put.yml:1]: Sure, adding more YAML tests for put

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorIndexServiceTests.java [hunk 8]
```diff
@@ -37,8 +41,73 @@ public void testPutConnector() throws Exception {
...
+        String connectorIdToDelete = connectorIds.get(0);
+        DeleteResponse resp = awaitDeleteConnector(connectorIdToDelete);
+        assertThat(resp.status(), equalTo(RestStatus.OK));
+        expectThrows(ResourceNotFoundException.class, () -> awaitGetConnector(connectorIdToDelete));
 
-        // TODO: more checks once GET endpoint is implemented :)
+        expectThrows(ResourceNotFoundException.class, () -> awaitDeleteConnector(connectorIdToDelete));
+    }
+
+    // TODO: figure out how to use ConnectorTemplateRegistry in the test node
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorIndexServiceTests.java:64]: Given that `listConnectors` endpoint is already comprehensively tested in YAML tests I would prefer to remove this test for this PR. 

The issue is that I'm so far unsuccessful at registering connector index templates (with default mappings) therefore this test fails for the list operation, since it expects the `connector_id` to be an indexed field.

I will ask/look around to get the templates registered for such a scenario, I couldn't find examples in the codebase.

**Comment by jedrazb** [comment]
### Changes
- Update `connector.*.json` with suggested description and feature flag visibility
- Add more comprehensive YAML tests for connector PUT test case
- Add JavaDocs to connector-related classes
- Split `ConnectorFiltering` class into separate files nested under `connector.filtering` package
- Add missing BWC tests for list/delete actions
- Answered all comments

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java [hunk 9]
```diff
@@ -172,25 +185,179 @@ public Connector(StreamInput in) throws IOException {
...
-    private static final ParseField LANGUAGE_FIELD = new ParseField("language");
-
-    private static final ParseField NAME_FIELD = new ParseField("name");
-    private static final ParseField PIPELINE_FIELD = new ParseField("pipeline");
-    private static final ParseField SCHEDULING_FIELD = new ParseField("scheduling");
-    private static final ParseField SERVICE_TYPE_FIELD = new ParseField("service_type");
-    private static final ParseField STATUS_FIELD = new ParseField("status");
-    private static final ParseField SYNC_CURSOR_FIELD = new ParseField("sync_cursor");
-    private static final ParseField SYNC_NOW_FIELD = new ParseField("sync_now");
+    protected static final ParseField ID_FIELD = new ParseField("connector_id");
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:188]: Why the change to protected?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java [hunk 10]
```diff
@@ -52,4 +71,134 @@ public void putConnector(Connector connector, ActionListener<DocWriteResponse> l
...
+            .map(ConnectorIndexService::hitToConnector)
+            .toList();
+        return new ConnectorIndexService.ConnectorResult(connectorResults, (int) response.getHits().getTotalHits().value);
+    }
+
+    private static Connector hitToConnector(SearchHit searchHit) {
+
+        // todo: don't return sensitive data from configuration in list endpoint
+
+        return Connector.fromXContentBytes(searchHit.getSourceRef(), XContentType.JSON);
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:169]: Related to our previous conversation, is this something that needs to be done, or can we just return the source directly without parsing it?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java [hunk 11]
```diff
@@ -96,17 +96,17 @@ public ConnectorSyncInfo(StreamInput in) throws IOException {
...
-    private static final ParseField LAST_ACCESS_CONTROL_SYNC_SCHEDULED_AT_FIELD = new ParseField("last_access_control_sync_scheduled_at");
-    private static final ParseField LAST_DELETED_DOCUMENT_COUNT_FIELD = new ParseField("last_deleted_document_count");
-    private static final ParseField LAST_INCREMENTAL_SYNC_SCHEDULED_AT_FIELD = new ParseField("last_incremental_sync_scheduled_at");
-    private static final ParseField LAST_INDEXED_DOCUMENT_COUNT_FIELD = new ParseField("last_indexed_document_count");
-    private static final ParseField LAST_SEEN_FIELD = new ParseField("last_seen");
-    private static final ParseField LAST_SYNC_ERROR_FIELD = new ParseField("last_sync_error");
-    private static final ParseField LAST_SYNC_SCHEDULED_AT_FIELD = new ParseField("last_sync_scheduled_at");
-    private static final ParseField LAST_SYNC_STATUS_FIELD = new ParseField("last_sync_status");
-    private static final ParseField LAST_SYNCED_FIELD = new ParseField("last_synced");
+    protected static final ParseField LAST_ACCESS_CONTROL_SYNC_ERROR = new ParseField("last_access_control_sync_error");
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java:99]: Same question, why the change to protected?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+public class DeleteConnectorAction extends ActionType<AcknowledgedResponse> {
+
+    public static final DeleteConnectorAction INSTANCE = new DeleteConnectorAction();
+    public static final String NAME = "cluster:admin/xpack/connector/delete";
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java:32]: Question: Are we having a new cluster admin privilege for these actions?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringRuleCondition.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package org.elasticsearch.xpack.application.connector.filtering;
+
+public enum FilteringRuleCondition {
+    CONTAINS("contains"),
+    ENDS_WITH("ends_with"),
+    EQUALS("equals"),
+    GT(">"),
+    LT("<"),
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringRuleCondition.java:15]: Nit: For no other reason than to use words on everything:

```suggestion
    GT("gt"),
    LT("lt"),
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringValidation.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+/**
+ * Represents the details of a validation process, including identifiers and descriptive messages.
+ * This class is used to encapsulate information about specific validation checks, where each validation
+ * is associated with a list of IDs and corresponding messages that detail the validation results.
+ */
+public class FilteringValidation implements Writeable, ToXContentObject {
+    private final List<String> ids;
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringValidation.java:31]: Are the messages related to the IDs? If so, shouldn't this be a List of tuples <id, message>?

**Review by carlosdelest — COMMENTED** [review]
Some comments - I'll try to do a second pass on this.

Adding individual operations would have make this PR (or the following ones) easier to review - please consider that for future API PRs

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringRuleCondition.java (hunk 13 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringRuleCondition.java:15]: Our protocol actually expects the `"<"` and `">"` in the doc so this is a workaround to support it. See reference [code in kibana](https://github.com/elastic/kibana/blob/main/packages/kbn-search-connectors/types/connectors.ts#L120)

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java (hunk 12 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java:32]: This is not a system index. Do we need any extra privileges? I was able to execute delete actions locally

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java (hunk 11 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java:99]: In `ConnectorIndexService` I support sorting by:
```
.sort(Connector.ID_FIELD.getPreferredName(), SortOrder.ASC);
``` 
So I need to have access to this field from a different class. 

For `update_x` actions I would need access to most of the parse fields in the connector doc. I decided to change all of the to protected for convenience.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java (hunk 10 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:169]: There was a request to obfuscate some user-provided secrets stored in the index (e.g. api keys to authenticate against a datasource). 

Given that - I needed to implement parses for all of "helper" connector data structures (for parsing update endpoints) - I decided that it might be elegant to:
1. Parse the whole connector document
2. Apply some internal function to obfuscate user secrets before returning data to the user

Note: the obfuscation logic is not implemented yet

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java (hunk 9 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:188]: In `ConnectorIndexService` I support sorting by:
```
.sort(Connector.ID_FIELD.getPreferredName(), SortOrder.ASC);
``` 
So I need to have access to this field from a different class. 

For `update_x` actions I would need access to most of the parse fields in the connector doc. I decided to change all of the to protected for convenience.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java (hunk 12 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java:32]: What are the permissions that a user need to have in order to manage connectors? Just permissions over the connectors index?

We probably want from a security / user roles perspective to limit those actions to specific users. These should probably be included as part of a `manage_connectors` cluster privilege.

Otherwise, we leave the option for non-authorized users to manage connectors and syncs.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringRuleCondition.java (hunk 13 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringRuleCondition.java:15]: Thanks for explaining!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java (hunk 10 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:169]: Got it, thanks!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java (hunk 9 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:188]: Thanks for explaining! If it's used by other classes in the package, it's better to use package protected access (no access modifier).

`protected` should be used just for subclass access. It can be confusing if subclasses are not using it or the class is not meant to be subclassed.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringValidation.java (hunk 14 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringValidation.java:31]: I doubt users would want to manually edit the basic/ advanced filtering rules by tinkering with raw JSON 😄 

Right now it's used by Kibana to manage the state of drafting and creation of new filtering rules for a connector. User will see this data when querying for the connector object. 

Not pasting an example directly here, but I'm linking the test from this PR that checks for data shape for connector filtering https://github.com/elastic/elasticsearch/pull/102592/files#diff-068f25a3a77ae7d85f0b826d1589f8cad2556673d509c845858e3a8bc3d803ab

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java (hunk 9 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/Connector.java:188]: Thank you for explanation :) I understand and I'm changing the access modifier then

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java (hunk 12 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java:32]: @carlosdelest Thank you for explaining, let me discuss this with the team. Do you think that adding the specific cluster privileges can be added in a separate PR? I need to figure out how to do this correctly.

File: rest-api-spec/src/main/resources/rest-api-spec/api/connector.delete.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (timgrein) [reply rest-api-spec/src/main/resources/rest-api-spec/api/connector.delete.json:4]: I wanted to add a comment that we should add a TODO until I've remembered again that JSON is data-only. I've created a corresponding issue and added it to the epic, so we don't forget this :)

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringValidation.java (hunk 14 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/filtering/FilteringValidation.java:31]: That could be changed for the user perspective so the two are aligned and remove the technical debt from the end users. But I'm ok with this - it's your domain and the API you're exposing, so it's your decision here 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java (hunk 11 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorSyncInfo.java]: But I guess that those fields in Connector are not related to this class, right? We're merely reusing them. I'd say in that case that we copy the fields in the Connector class, as they are not related.

**Review by carlosdelest — APPROVED** [review]
Good job! 💯 

My only concern is about security, and that can be addressed in a separate PR.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java (hunk 12 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/DeleteConnectorAction.java:32]: Thank you for sharing context, if it's indeed a small change I will try to squeeze this before the FF :)

File: rest-api-spec/src/main/resources/rest-api-spec/api/connector.delete.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (kderusso) [reply rest-api-spec/src/main/resources/rest-api-spec/api/connector.delete.json:4]: Awesome, thank you! 
Note: We could consider placeholder stubs as well, but this is fine for now.

**Review by kderusso — APPROVED** [review]
Thanks for iterating on this @jedrazb , very nice work. Looks great. The class comments are great.
